feat(config): add smart-ret for corfu

LemonBreezes opened the pull request:
Hi. This re-adds that change where RET will quit Corfu and pass-through if no candidate is selected. It's particularly ergonomic if you set a small `corfu-auto-delay` because then Corfu is popping up constantly as you type and you keep having to type RET twice.

LemonBreezes committed:
feat(config): add smart-ret for corfu

This commit makes RET in Corfu quit auto-completion and passthrough to
the underlying keymap if no completion is selected.

LuigiPiucco commented:
Hi. Just before merging, does it make sense to gate this addition behind `+tng`? For the opposite case, it feels strange to do this, since it's the only way we can confirm, but `+tng` has TAB to just insert. Also, how would unbinding RET compare in ergonomics? Do you use it in addition to TAB or is it just a bother?

LemonBreezes commented:
> Hi. Just before merging, does it make sense to gate this addition behind `+tng`? For the opposite case, it feels strange to do this, since it's the only way we can confirm, but `+tng` has TAB to just insert. Also, how would unbinding RET compare in ergonomics? Do you use it in addition to TAB or is it just a bother?

Even if you use `+tng`, you need RET for some backends like `yasnippet-capf`. I think it's a good default because the completion popup should not affect your ordinary typing if you're not using it.

LemonBreezes committed:
fix(corfu): move orderless conf out of corfu block

LemonBreezes commented:
I also moved the orderless separator config outside of the `(use-package! corfu ...)` block.

LemonBreezes commented:
I just realized this DWIM-style return option is also in `evil-collection-corfu` along with our backspace hack. Though of course, that package depends on Evil.

![image](https://github.com/LuigiPiucco/doom-emacs/assets/29102529/7f1e47f0-f761-4727-b6ca-1c294bbfdf38)

LemonBreezes committed:
fix(corfu): complete after colon in lispy

LemonBreezes committed:
fix(corfu): use :filter instead of :enable

LemonBreezes committed:
fix(corfu): improve detection of comments

Due to https://github.com/minad/cape/pull/109, we cannot rely on only
faces to detect comments.

LemonBreezes committed:
fix(corfu): eval lambdas

LemonBreezes committed:
feat(config): modify smart-ret in +tng minibuffer

If we are using +tng, we can make RET always exit the minibuffer. This
has the downside of not being able to insert snippets, but no one writes
snippets for the minibuffer anyway.

LemonBreezes committed:
feat(corfu): make minibuffer completion optional

LuigiPiucco commented:
First, sorry for the wait. College can be very consuming from time to time.

I really like 99% of the changes here, and thank you for putting this much effort in, but the smart-ret commit still feels slightly strange to me. To check if I understand the purpose: when typing, you are bothered by the fact that completion can popup anywhere and prevent you from (for instance) breaking a line. Now, there is something to note here: when you type RET and have this issue, does Corfu insert the completion or just quit? The variable `corfu-preselect` controls this indirectly. When set to `prompt`, it should just quit, and that's the behavior of `+tng` currently. I use non-`+tng`, so in this situation, the first candidate is preselected, therefore I get what I expect, a completion and no newline. If I had the prompt preselected, however, I'd expect the newline to be inserted. Does this check out with you? If yes, I believe it may be worth asking @minad if there's already a way to do this, or whether this could be implemented directly in Corfu. For reference, something like the following, maybe as an alternate insert function:

```emacs-lisp
(defun corfu-insert (&optional prefix) ;; Either replace or implement as alternative
  "Insert current candidate.
Quit if no candidate is selected.

If no candidate is selected and PREFIX is non-nil, check the key vector that
caused this command to be called and look it up after the popup quits. If bound,
call that binding."
  (interactive "p")
  (if (>= corfu--index 0)
      (corfu--insert 'finished)
    (corfu-quit)
    (when-let ((prefix)
               (cmd (keymap-lookup nil (key-description (this-command-keys-vector)))))
      (call-interactively cmd))))
```
When bound to `RET`, for instance, this passes-through to the newline if prompt is selected (or quits the minibuffer without completing, if it is active). When a candidate is chosen via other bindings or via `corfu-preselect`, it only inserts the candidate.

I think this is general enough to implement directly in Corfu, and could facilitate its use in the minibuffer as well. It becomes unobtrusive when `corfu-preselect` is `prompt`, which maybe we should make the default for non-`+tng` as well. Thus, it would make the whole `RET` vs. `C-RET` vs. `S-RET` thing unnecessary. Also seems cleaner to me than the current menu-item approach.

For Daniel, I can PR such a change if you accept it.

minad commented:
@LuigiPiucco It is not clear what the ideal "smart" RET behavior looks like. I think it is even good to disable RET (except in shells) as I've documented in https://github.com/minad/corfu?tab=readme-ov-file#tab-only-completion. This is what I use right now.

LuigiPiucco commented:
> @LuigiPiucco It is not clear what the ideal "smart" RET behavior looks like. I think it is even good to disable RET (except in shells) as I've documented in https://github.com/minad/corfu?tab=readme-ov-file#tab-only-completion. This is what I use right now.

I see. Just to make sure, you think implementing more alternatives in base Corfu (`corfu-insert`, `corfu-insert-or-passthrough` and leaving RET for the user to bind/unbind) doesn't bring value, also?

I'll wait for @LemonBreezes's opinion before proceeding. Did I at least understand correctly what is being fixed here? Maybe I'm overcomplicating things.

minad commented:
> I see. Just to make sure, you think implementing more alternatives in base Corfu (corfu-insert, corfu-insert-or-passthrough and leaving RET for the user to bind/unbind) doesn't bring value, also?

Not right now. I don't want to increase the number of usage variants. There are already plenty of possibilities.

Also I think it would be good to focus on getting the PR merged instead of optimizing details such that Corfu becomes available to a wider user base. This way you may also get additional feedback.

LemonBreezes commented:
> First, sorry for the wait. College can be very consuming from time to time.
> 
> I really like 99% of the changes here, and thank you for putting this much effort in, but the smart-ret commit still feels slightly strange to me. To check if I understand the purpose: when typing, you are bothered by the fact that completion can popup anywhere and prevent you from (for instance) breaking a line. Now, there is something to note here: when you type RET and have this issue, does Corfu insert the completion or just quit? The variable `corfu-preselect` controls this indirectly. When set to `prompt`, it should just quit, and that's the behavior of `+tng` currently. I use non-`+tng`, so in this situation, the first candidate is preselected, therefore I get what I expect, a completion and no newline. If I had the prompt preselected, however, I'd expect the newline to be inserted. Does this check out with you? If yes, I believe it may be worth asking @minad if there's already a way to do this, or whether this could be implemented directly in Corfu. For reference, something like the following, maybe as an alternate insert function:
> 
> ```emacs-lisp
> (defun corfu-insert (&optional prefix) ;; Either replace or implement as alternative
>   "Insert current candidate.
> Quit if no candidate is selected.
> 
> If no candidate is selected and PREFIX is non-nil, check the key vector that
> caused this command to be called and look it up after the popup quits. If bound,
> call that binding."
>   (interactive "p")
>   (if (>= corfu--index 0)
>       (corfu--insert 'finished)
>     (corfu-quit)
>     (when-let ((prefix)
>                (cmd (keymap-lookup nil (key-description (this-command-keys-vector)))))
>       (call-interactively cmd))))
> ```
> 
> When bound to `RET`, for instance, this passes-through to the newline if prompt is selected (or quits the minibuffer without completing, if it is active). When a candidate is chosen via other bindings or via `corfu-preselect`, it only inserts the candidate.
> 
> I think this is general enough to implement directly in Corfu, and could facilitate its use in the minibuffer as well. It becomes unobtrusive when `corfu-preselect` is `prompt`, which maybe we should make the default for non-`+tng` as well. Thus, it would make the whole `RET` vs. `C-RET` vs. `S-RET` thing unnecessary. Also seems cleaner to me than the current menu-item approach.
> 
> For Daniel, I can PR such a change if you accept it.

Making a command like that which calls other commands with `keymap-lookup` is generally a bad idea. The reason is that it does not take into account `overriding-terminal-local-map`, `key-translation-map`, `<return>` commands falling back to `RET`, etc. To implement the full keymap lookup logic in Elisp would be extremely difficult. There are two currently suggested ways to do what we want (https://debbugs.gnu.org/cgi/bugreport.cgi?bug=66755):
1. Using the `:filter` menu-item. 
2. Take `this-single-command-raw-keys` and push it back on `unread-command-events`. 

Of these, 1. is the most robust and efficient. I might make another smaller PR this weekend.

LuigiPiucco commented:
Ok, I probably was overcomplicating things. I will use the menu-item based implementation of this. Also, I think it could be good to allow the user to enable features individually, rather than just `+tng` vs. not. I'll push a prototype later, that hopefully is ok for all of us. If not, it will be in the form of additional commits, so we can rebase them out.
